_pages/dev/tms.md

```diff
@@ -0,0 +1,243 @@
+
+If you use a custom data layer object (with the standard data structure), you can configure it using the `dataLayerProperty` configuration property.
+
+If you are using a different data layer structure than the standard, you can write your own collector like this:
+
+```typescript
+import { Injectable } from "@angular/core";
+import { CxEvent } from "@spartacus/core";
+import {
+  TmsCollector,
+  TmsCollectorConfig,
+  WindowObject,
+} from "@spartacus/tracking/tms/core";
+
+@Injectable({ providedIn: "root" })
+export class MyCollectorService implements TmsCollector {
+  init(config: TmsCollectorConfig, windowObject: WindowObject): void {
+    windowObject.myDataLayer = windowObject.myDataLayer ?? [];
+  }
+
+  pushEvent<T extends CxEvent>(
+    config: TmsCollectorConfig,
+    windowObject: WindowObject,
+    event: T | any
+  ): void {
+    windowObject.myDataLayer.push(event);
+  }
+}
+```
+
+and configure it like this:
+
+```typescript
+BaseTmsModule.forRoot({
+  tagManager: {
+    gtm: {
+      collector: MyCollectorService,
+      events: [NavigationEvent, CartAddEntrySuccessEvent],
+    },
+    ...
+  },
+}),
+```
+
+#### Events payload
+
+As the required event's payloads can quite vary between clients, we can't cover every possible use case. With that in mind, we've made sure that the events have some sensible payloads, and we've made it easy to adjust (re-map) the payloads in more than one way:
+
+##### Creating a custom mapper
+
+To create a custom mapper per a TMS solution, you can build on top of the [example above](#Data-layer) (by creating a custom collector), and implement the `map` method like so:
+
+```typescript
+import { Injectable } from "@angular/core";
+import { CxEvent } from "@spartacus/core";
+import {
+  TmsCollector,
+  TmsCollectorConfig,
+  WindowObject,
+} from "@spartacus/tracking/tms/core";
+
+@Injectable({ providedIn: "root" })
+export class MyCollectorService implements TmsCollector {
+  ...
+
+  map?<T extends CxEvent>(event: T): T | object {
+    // TODO: custom mapping logic goes here...
+  }
+}
+```
+
+In the `map` method you can use `instanceof` check to differentiate which event should you re-map. As this logic can get quite complex, you have the ability to inject other services to your `MyCollectorService`.
+
+##### Creating custom events
+
+Creating custom events in Spartacus and re-mapping the data to fit your data structure requirements is straightforward:
+
+```typescript
+import { Injectable } from "@angular/core";
+import { createFrom, EventService } from "@spartacus/core";
+import { NavigationEvent } from "@spartacus/storefront";
+import { Observable } from "rxjs";
+import { map } from "rxjs/operators";
+
+export class CustomNavigationEvent {
+  pageType: string;
+  pageName: string;
+}
+
+@Injectable({
+  providedIn: "root",
+})
+export class CustomNavigationEventBuilder {
+  constructor(protected eventService: EventService) {
+    this.buildNavigationEvent();
```

Comment:
This example doesn't look complete IMO.

Also, you don't need a concept of a builder (we introduced it to map actions to events). 

In case of events you just have to create new event source, it can compose other events like this example:
```
const eventSource = this.eventService.get(NavigationEvent).pipe(
      map((navigationEvent) =>
        createFrom(CustomNavigationEvent, {
          // just an example logic, it's not necessarily the actual page name and type
          pageName: navigationEvent.context.id,
          pageType: navigationEvent.context.type,
        })
      )
      ```
and register this source with eventService.register()
(and it's described in the event service, so maybe referencing it is a good idea)

Reply:
Good point, done